tests/examples/components/shared/test_aml_comm.py

```diff
@@ -0,0 +1,181 @@
+
+
+# Mock the redis connection to avoid having to run a redis server
+class TestAMLCommRedis(AMLCommRedis):
+    def __init__(self, buffer, *args, **kwargs):
+        self.buffer = buffer
+        self._get_connection_string = MagicMock(return_value="localhost")
+        self._format_connection_string = MagicMock(return_value="localhost")
+        self._wait_for_connection = MagicMock()
+        self._connect_to_redis = MagicMock()
+        super().__init__(*args, **kwargs)
+        self._client = None
+
+    def _get_session_id(self, source, destination, flag=None):
+        session_id = super()._get_session_id(source, destination, flag)
+        if session_id not in self.buffer:
+            self.buffer[session_id] = []
+        return session_id
+
+    def _send(self, data, session_id, _) -> None:
+        self.buffer[session_id] = self.buffer[session_id] + [data]
+
+    def _recv(self, session_id, _) -> bytes:
+        while len(self.buffer[session_id]) == 0:
+            time.sleep(1)
+        item = self.buffer[session_id][0]
+        self.buffer[session_id] = self.buffer[session_id][1:]
+        return item
+
+
+class TestAMLComm(unittest.TestCase):
+    def test_aml_comm_socket_simple(self):
+        TEST_MSG = lambda recv: f"Message to {recv}"
+
+        for encrypted in [True, False]:
+            for use_redis in [True, False]:
+                with self.subTest(encrypted=encrypted, use_redis=use_redis):
+                    # Create two processes that send each other message
+                    manager = mp.Manager()
+                    shared_dict = manager.dict()
+                    port = get_free_port()
+                    p1 = mp.Process(
+                        target=init_send_recv,
+                        args=(
+                            0,
+                            2,
+                            "test_run_id",
+                            "localhost",
+                            port,
+                            TEST_MSG(1),
+                            TEST_MSG(0),
+                            shared_dict,
+                            encrypted,
+                            use_redis,
+                        ),
+                    )
+                    p2 = mp.Process(
+                        target=init_send_recv,
+                        args=(
+                            1,
+                            2,
+                            "test_run_id",
+                            "localhost",
+                            port,
+                            TEST_MSG(0),
+                            TEST_MSG(1),
+                            shared_dict,
+                            encrypted,
+                            use_redis,
+                        ),
+                    )
+                    p1.start()
+                    p2.start()
+                    p1.join()
+                    p2.join()
+
+                    for i in range(2):
+                        self.assertTrue(shared_dict[i])
```

Comment:
Maybe log the test results explicitly?

Reply:
What would be reason for that? Afaik. this is regular assertion you make in unittest.

Reply:
I'm not familiar with the unittest framework. I suggested that because I'm just not sure of what exactly is surfaced when a subtest fails - does it clearly expose the value of the `encrypted` and `use_redis` parameters? If so, then disregard. If not, then knowing exactly which subtest failed would be useful.